group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -544,6 +579,100 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchAllOffsets(
             .setTopics(topicResponses);
     }
 
+    /**
+     * Remove expired offsets for group.
+     *
+     * @param groupId The group id.
+     * @param records The list of records to populate with offset commit tombstone records.
+     *
+     * @return True if no offsets exist or if all offsets expired, false otherwise.
+     */
+    public boolean cleanupExpiredOffsets(String groupId, List<Record> records) {
+        TimelineHashMap<String, TimelineHashMap<Integer, OffsetAndMetadata>> offsetsByTopic = offsetsByGroup.get(groupId);
+        if (offsetsByTopic == null) {
+            return true;
+        }
+
+        // We expect the group to exist.
+        Group group = groupMetadataManager.group(groupId);
+        Set<TopicPartition> expiredPartitions = new HashSet<>();
+        long currentTimestamp = time.milliseconds();
+        Optional<OffsetExpirationCondition> offsetExpirationCondition = group.offsetExpirationCondition();
+
+        if (!offsetExpirationCondition.isPresent()) {
+            return false;
+        }
+
+        AtomicBoolean hasAllOffsetsExpired = new AtomicBoolean(true);
+        OffsetExpirationCondition condition = offsetExpirationCondition.get();
+
+        offsetsByTopic.forEach((topic, partitions) -> {
+            if (!group.isSubscribedToTopic(topic, false)) {
+                partitions.forEach((partition, offsetAndMetadata) -> {
+                    if (condition.isOffsetExpired(offsetAndMetadata, currentTimestamp, config.offsetsRetentionMs)) {
+                        expiredPartitions.add(appendOffsetCommitTombstone(groupId, topic, partition, records));
+                    } else {
+                        hasAllOffsetsExpired.set(false);
+                    }
+                });
+            } else {
+                hasAllOffsetsExpired.set(false);
+            }
+        });
+
+        log.debug("[GroupId {}] Expiring offsets: {}", groupId, expiredPartitions);
```

Comment:
I wonder if we should log this with info level. What do you think? We should also format `expiredPartitions` correctly here.

Reply:
changed to info and to only log when there are expired partitions.

What do you mean by format, as in just log the size?

Reply:
I meant that we need to format the list to a comma separated string. Could we also log the boolean indicating whether all the offsets get expired or not.

Should we also say `Expiring offsets of partitions: `?